.github/workflows/build.yaml

```diff
@@ -66,33 +66,79 @@ jobs:
           provenance: false
           push: true
           tags: |
-            ${{ needs.setup.outputs.image_repository }}:latest-${{ steps.set-build-variables.outputs.PLATFORM_OS }}-${{ steps.set-build-variables.outputs.PLATFORM_ARCH }}
-            ${{ needs.setup.outputs.image_repository }}:${{ needs.setup.outputs.image_tag }}-${{ steps.set-build-variables.outputs.PLATFORM_OS }}-${{ steps.set-build-variables.outputs.PLATFORM_ARCH }}
+            ${{ env.MAIN_REGISTRY }}/${{ env.MAIN_REPOSITORY }}:latest-${{ steps.set-build-variables.outputs.PLATFORM_DASH }}
+            ${{ env.MAIN_REGISTRY }}/${{ env.MAIN_REPOSITORY }}:${{ needs.setup.outputs.image_tag }}-${{ steps.set-build-variables.outputs.PLATFORM_DASH }}
 
-  process-image-manifest:
+  push-images-and-manifests:
     runs-on: ubuntu-latest
-    name: process-image-manifest
-    needs: ["setup", "build"]
+    needs: ["setup", "build-images"]
+    strategy:
+      matrix:
+        include:
+          - registry: ghcr.io
+            username: GHCR_USERNAME
+            password: GITHUB_TOKEN
+            repository: ${{ github.repository }}
+          # Define more image repositories like this:
+          # - registry: quay.io
+          #   username: QUAY_USERNAME
+          #   password: QUAY_TOKEN
+          #   repository: janus/redhat-backstage-build
+
     steps:
-      - name: Log in to Registry
+      - name: Log in to Registry to pull images
+        uses: redhat-actions/podman-login@632d91dfe19e1b55833cb708786bfbad2c2a0335 #pin@v3.2.0
+        with:
+          registry: ${{ env.MAIN_REGISTRY }}
+          username: ${{ vars[env.MAIN_USERNAME] }}
+          password: ${{ secrets[env.MAIN_PASSWORD] }}
+
+      - name: Import Docker images
+        run: |
+          for PLATFORM in ${{ needs.setup.outputs.image_platform_tags }}; do \
+            PLATFORM_SLASH=$(echo $PLATFORM | sed -e 's/-/\//g')
+            for TAG in latest ${{ needs.setup.outputs.image_tag }}; do \
+              docker pull --platform $PLATFORM_SLASH ${{ env.MAIN_REGISTRY }}/${{ env.MAIN_REPOSITORY }}:$TAG-$PLATFORM
+            done
+          done
+
+      - name: Log out of registry
+        shell: bash
+        run: |
+          podman logout ${{ env.MAIN_REGISTRY }}
+
+      - name: Log in to Registry to push images
         uses: redhat-actions/podman-login@632d91dfe19e1b55833cb708786bfbad2c2a0335 #pin@v3.2.0
         with:
-          username: ${{ github.actor }}
-          password: ${{ secrets.GITHUB_TOKEN }}
-          registry: ${{ needs.setup.outputs.image_registry }}
-      - name: "Append to Bundle Image Manifest"
+          registry: ${{ matrix.registry }}
+          username: ${{ vars[matrix.username] }}
+          password: ${{ secrets[matrix.password] }}
```

Comment:
Can we please expand these vars directly in the `matrix.include`? That way we can use references like `${{ github.actor }} for the username etc.

The current implementation expects user to always know that username is part of `vars`, password is in `secrets` etc. Which may not be always obvious.

Reply:
Secrets can't be included in the matrix generation, otherwise they would be revealed. So when I try to put `${{ secrets.SOME_TOKEN }}` straight into the matrix definition, the wokflow isn't even ran. But `vars` work I'll just move them to the matrix definition.